services/transactionService.ts

```diff
@@ -0,0 +1,249 @@
+import { errorMessages } from '../utils/errorMessages';
+import {
+  NetworkTransactionInfo,
+  TransactionDetailInput,
+} from '../types/TransactionInquiry';
+import axios from 'axios';
+import { erc20ABI } from '../assets/erc20ABI';
+import {
+  getEtherscanOrBlockScoutUrl,
+  getNetworkNativeToken,
+  getNetworkWeb3,
+} from '../provider';
+
+abiDecoder.addABI(erc20ABI);
+
+export async function getTransactionInfoFromNetwork(
+  input: TransactionDetailInput,
+): Promise<NetworkTransactionInfo> {
+  const { networkId, nonce } = input;
+
+  const web3 = getNetworkWeb3(networkId);
+  const userTransactionsCount = await web3.eth.getTransactionCount(
+    input.fromAddress,
+  );
+  if (nonce && userTransactionsCount <= nonce) {
+    console.log('getTransactionDetail check nonce', {
+      input,
+      userTransactionsCount,
+    });
+    throw new Error(
+      errorMessages.TRANSACTION_WITH_THIS_NONCE_IS_NOT_MINED_ALREADY,
+    );
+  }
+  let transaction: NetworkTransactionInfo | null = await findTransactionByHash(
+    input,
+  );
+
+  if (!transaction && nonce) {
+    // if nonce didn't pass, we can not understand whether is speedup or not
+    transaction = await findTransactionByNonce({
+      input,
+    });
+  }
+  if (!transaction) {
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUND);
+  }
+  validateTransactionWithInputData(transaction, input);
+  return transaction;
+}
+
+async function findTransactionByHash(input: TransactionDetailInput) {
+  const nativeToken = getNetworkNativeToken(input.networkId);
+  if (nativeToken === input.symbol) {
+    return getTransactionDetailForNormalTransfer(input);
+  } else {
+    return getTransactionDetailForTokenTransfer(input);
+  }
+}
+
+async function findTransactionByNonce(data: {
+  input: TransactionDetailInput;
+  page?: number;
+}): Promise<NetworkTransactionInfo | null> {
+  console.log('findTransactionByNonce called', data);
+  const { input, page = 1 } = data;
+  const nonce = input.nonce as number;
+  const { userRecentTransactions, isTransactionListEmpty } =
+    await getListOfTransactionsByAddress({
+      address: input.fromAddress,
+      page,
+      networkId: input.networkId,
+    });
+  if (isTransactionListEmpty) {
+    // we know that we reached to end of transactions
+    console.log('findTransactionByNonce, no more found donations for address', {
+      page,
+      address: input.fromAddress,
+    });
+    throw new Error(errorMessages.TRANSACTION_NOT_FOUND);
+  }
+  const foundTransaction = userRecentTransactions.find(
+    tx => tx.nonce === input.nonce,
+  );
+
+  if (foundTransaction) {
+    return findTransactionByHash({ ...input, txHash: foundTransaction.hash });
+  }
+
+  // userRecentTransactions just includes the transactions that source is our fromAddress
+  // so if the lowest nonce in this array is smaller than the sent nonce we would know that we should not
+  // check latest transactions
+  const smallestNonce: number = userRecentTransactions[
```

Comment:
userRecentTransactions length can be zero

Reply:
it has been fixed in #185